editor/src/clj/editor/font.clj

```diff
@@ -160,46 +160,119 @@
+        len (.length line)]
+    (if-let [last (get glyphs (and (pos? len) (int (.charAt line (dec len)))))]
       (- (+ w (:left-bearing last) (:width last)) (:advance last))
       w)))
 
-(defn- split [s i]
-  (mapv s/trim [(subs s 0 i) (subs s i)]))
-
-(defn- break-line [^String line glyphs max-width text-tracking]
-  (loop [b (dec (count line))
-         last-space nil]
-    (if (> b 0)
-      (let [c (.charAt line b)
-            cu (.codePointAt line b)]
-        (if (or (Character/isWhitespace c) (= cu 0x200B))
-          (let [[l1 l2] (split line b)
-                w (measure-line glyphs text-tracking l1)]
-            (if (> w max-width)
-              (recur (dec b) b)
-              [l1 l2]))
-          (recur (dec b) last-space)))
-      (if last-space
-        (split line last-space)
-        [line nil]))))
-
-(defn- break-lines [lines glyphs max-width text-tracking]
-  (reduce (fn [result line]
-            (let [trimmed-line (s/trim line)
-                  w (measure-line glyphs text-tracking trimmed-line)]
-              (if (> w max-width)
-                (let [[l1 l2] (break-line trimmed-line glyphs max-width text-tracking)]
-                  (if l2
-                    (recur (conj result l1) l2)
-                    (conj result trimmed-line)))
-                (conj result trimmed-line)))) [] lines))
-
-(defn- split-text [glyphs text line-break? max-width text-tracking]
-  (cond-> (s/split-lines text)
-    line-break? (break-lines glyphs max-width text-tracking)))
+(defn- split-text [glyphs ^String text line-break? max-width text-tracking]
+  (if line-break?
+    ;; Rules for line breaks:
+    ;; 1. if a single line without spaces exceeds max width, let it be so
+    ;; 2. always split lines on \r?\n
+    ;; 3. remove trailing empty lines
+    ;; 4. trim start and end of a single line
+    ;; 5. if a line is too long, try splitting it on whitespace (including zero-width
+    ;;    space)
+    (let [text-length (.length text)
+          last-index (dec text-length)
+          zero-width-space (char 0x200B)
+          ;; remove trailing whitespace from sb and add to lines
+          add-line! (fn [lines ^StringBuilder sb]
+                      (let [line-length (.length sb)]
+                        (if (zero? line-length)
+                          (conj! lines "")
+                          (let [last-non-ws-index
+                                ;; Note: normally, this loop would be a subject to IOOB
+                                ;; exception, but here we have an invariant that
+                                ;; the accumulated string in sb is either empty or does
+                                ;; not start with a white space char
+                                (loop [i (dec line-length)]
+                                  (let [ch (.charAt sb i)]
+                                    (if (or (Character/isWhitespace ch)
+                                            (= zero-width-space ch))
```

Comment:
This `or` is repeated a couple of times in the code. I was thinking of moving it to a standalone function, but then we would have char boxing, which would hurt the performance.